appinventor/components/src/com/google/appinventor/components/runtime/SpeechRecognizer.java

```diff
@@ -103,11 +126,67 @@ public void BeforeGettingText() {
   }
 
   /**
-   * Simple event to raise after the VoiceReco activity has returned
+   * Simple event to raise after the VoiceReco activity has returned.
+   * Partial is 'false' when SpeechRecognizer stops automatically after listening,
+   * else it is 'true' if SpeechRecognizer returns partial results.
    */
   @SimpleEvent
-  public void AfterGettingText(String result) {
-    EventDispatcher.dispatchEvent(this, "AfterGettingText", result);
+  public void AfterGettingText(String result, boolean partial) {
+    EventDispatcher.dispatchEvent(this, "AfterGettingText", result, partial);
+  }
+
+  /**
+   * Method from SpeechListener interface.
+   */
+  @Override
+  public void onPartialResult(String text) {
+    result = text;
+    AfterGettingText(result, true);
+  }
+
+  /**
+   * Method from SpeechListener interface.
+   */
+  @Override
+  public void onResult(String text) {
+    result = text;
+    AfterGettingText(result, false);
+    onDelete();
+  }
+
+  /**
+   * Method from SpeechListener interface.
+   */
+  @Override
+  public void onError(String message) {
+    result = message;
+    AfterGettingText(result, false);
   }
 
+  @Override
+  public void onDelete() {
+    speechRecognizerController = null;
+    recognizerIntent = null;
+  }
+
+  @SimpleProperty(
+      category = PropertyCategory.BEHAVIOR,
+      description = "If true, an app can retain their older behaviour.")
+  public boolean UseLegacy() {
+    return useLegacy;
+  }
+
+  @DesignerProperty(editorType = PropertyTypeConstants.PROPERTY_TYPE_BOOLEAN,
+      defaultValue = "True")
+  @SimpleProperty(description = "If true, a separate dialog is used to recognize speech. "
+      + "If false, speech is recognized without changing the user interface and "
+      + "partial results are also provided.")
+  public void UseLegacy(boolean useLegacy) {
+    this.useLegacy = useLegacy;
+    if (useLegacy == true || Build.VERSION.SDK_INT<8) {
+      speechRecognizerController = new IntentBasedSpeechRecognizer(container, recognizerIntent);
```

Comment:
See my comment above about stopping before setting speechRecognizerController to something else. Otherwise people might be able to start multiple recognizers which is likely undesirable.

Reply:
Also, if we are avoiding creation of multiple instances of SpeechRecognizer here, then, along with stopping speechRecognizerController, shouldn't we also initialize it to 'null' before re-initializing with appropriate controller?

Reply:
I think, initializing to null is not required. but still just confirming once.
I was thinking about cases when people can start multiple recognizers. Will that happens when start button is clicked again while SpeechRecognizer is already listening (at the same moment)?

Reply:
True, it seems like you should have a flag to represent when speech is/isn't in progress so that you don't call through to the start method more than once. In the Intent-based case, I think it just brings forward the existing speech recognizer. It's unclear what happens in the Service-based case. It would be best just to avoid the ambiguity altogether.

Reply:
@ewpatton I have added Stop() method.
Also, regarding adding flag, it will not be required I guess. From what I checked,
Consider cases:
- Using IntentBasedSpeechRecognizer: Since another custom activity opens up for listening speech, if we try again to click on start button, this custom activity stops and hence listening stops.
- Using ServiceBasedSpeechRecognizer: If I try to click start button again, I will just get an error message "RecognitionService Busy"(already handled in ServiceBasedSpeechRecognizer class), which disappears in a second since recognizer is still listening, and no speech is lost. SpeechRecognizer will behave as if start button was never clicked.